race-buffer/src/writer.rs

```diff
@@ -0,0 +1,252 @@
+        Result<RaceBuffer<'a, E>, SizeError>,
+        &'a mut [u8],
+    ) {
+        // Safe because storage is treated as uninit after transmutation
+        let (prefix, storage, suffix) = unsafe { bytes.align_to_mut() };
+        (prefix, RaceBuffer::new(storage), suffix)
+    }
+
+    /// Write single entry to buffer, returning overwritten entry
+    pub fn write(&mut self, entry: E) -> OverwrittenEntry<E> {
+        let mut overwritten = OverwrittenEntry::None;
+        if self.wcurs == self.owcurs {
+            // Reading storage is safe after checking if buffer has looped at least once
+            if self.wcurs >= self.storage.len()
+                && unsafe { self.read_storage(self.wcurs).is_prefix() }
+            {
+                // If overwriting double, increment overwrite cursor by 2
+                overwritten = unsafe {
+                    OverwrittenEntry::Double(
+                        self.read_storage(self.owcurs),
+                        self.read_storage(self.owcurs + 1),
+                    )
+                };
+                self.owcurs += 2;
+            } else {
+                overwritten = unsafe { OverwrittenEntry::Single(self.read_storage(self.owcurs)) };
+                self.owcurs += 1;
+            }
+        }
+        self.write_to_storage(self.wcurs, entry);
+        self.wcurs += 1;
+        overwritten
+    }
+
+    /// Write double entry in single borrow, returning overwritten entry
+    pub fn write_double(
+        &mut self,
+        prefix: E,
+        suffix: E,
+    ) -> (OverwrittenEntry<E>, OverwrittenEntry<E>) {
+        let first_overwritten = self.write(prefix);
+        let second_overwritten = self.write(suffix);
+        (first_overwritten, second_overwritten)
+    }
+
+    /// Get value of backing storage corresponding to given index
+    #[inline]
+    pub(crate) unsafe fn read_storage(&self, curs: usize) -> E {
+        self.storage[get_cursor_index(self.storage.len(), curs, self.use_base_2_indexing)]
+            .assume_init()
+    }
+
+    #[inline]
+    fn write_to_storage(&mut self, curs: usize, entry: E) {
+        self.storage[get_cursor_index(self.storage.len(), curs, self.use_base_2_indexing)] =
+            MaybeUninit::new(entry);
+    }
+
+    #[inline]
+    pub fn read(&self, curs: usize) -> ReadEntry<E> {
+        if self.wcurs <= curs {
+            ReadEntry::NotYetWritten
+        } else if self.owcurs > curs + self.storage.len() {
+            ReadEntry::Missed
+        } else {
+            ReadEntry::Entry(unsafe {
+                self.read_storage(get_cursor_index(
+                    self.storage.len(),
+                    curs,
+                    self.use_base_2_indexing,
+                ))
+            })
+        }
+    }
+
+    #[inline]
+    pub fn iter<'b>(&'b self, start_rcurs: usize) -> RaceBufferIter<'a, 'b, E> {
+        RaceBufferIter::new(self, start_rcurs)
+    }
+
+    #[inline]
+    pub fn as_slice(&self) -> &[MaybeUninit<E>] {
```

Comment:
Seems odd to expose these internals as a public function.  Can we avoid this?

Reply:
Right now the `as_slice` function is used to get a pointer to the storage to ensure it isn't overlapping with the clocks list when creating a new DynamicHistory. Other than that, it isn't necessary